ctxutil.FastValue prototyping

RaduBerinde opened the pull request:
Epic: none
Release note: None

#### go.mod: update logtag dependency

Epic: none
Release note: None

#### go.mod: update crlib dependency

Epic: none
Release note: None

#### WIP


#### serverident-key


#### serverident-fast-value

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/136781)
<!-- Reviewable:end -->

RaduBerinde committed:
go.mod: update logtag dependency

See https://github.com/cockroachdb/logtags/pull/4

Epic: none
Release note: None

RaduBerinde committed:
serverident: unexport ServerIdentificationContextKey

Epic: none
Release note: None

RaduBerinde committed:
ctxutil: introduce fast values

This commit introduces a custom context type that can be used to more
efficiently associate values to contexts.

Instead of using arbitrary objects as keys, the keys are a small set
of integers. Any key must be globally registered first, and there is a
limited number that can be registered.

The `fastValuesCtx` stores *all* values for these keys, so traversing
the context hierarchy is not necessary. We also provide a way to add
multiple values at the same time, which saves on allocations.

I looked at before and after profiles in sysbench. Before:
 - CPU usage:
  context.value 0.6%
  context.WithValue 0.3%
  total 0.9%
 - Allocs:
  context.WithValue: 3.5%

After:
 - CPU usage:
  context.value + ctxutil.FastValue 0.5%
  context.WithValue 0.1%
  ctxutil.WithFastValue(s) 0.1%
  total 0.7%
 - Allocs:
  context.WithValue: 0.2%
  context.WithFastValue: 0.6%
  ctxutil.FastValuesBuilder.Finish: 1.4%
  total 2.2%

I will investigate improving on this, perhaps with providing a
pre-allocated context in codepaths where this is possible.

Informs: #136581
Release note: None
